File: InterPerter_Pattern/cpp/Interpreter_Pattern.cc
#include <iostream>
#include <vector>
#include <stdlib.h>
using namespace std;


class interpreter {
public:
    virtual int InterPreter() = 0;
};
int Option(vector<string>& v, vector<interpreter*>& iv);
class AddInterpreter : public interpreter {
public:
    AddInterpreter(interpreter* f, interpreter* s):frist(f),second(s){}
public:
    int InterPreter() {
        return frist->InterPreter() + second->InterPreter();
    }
private:
    interpreter* frist;
    interpreter* second;
};
class MutilInterpreter : public interpreter {
public:
    MutilInterpreter(interpreter* f, interpreter* s):frist(f),second(s){}
public:
    int InterPreter() {
        return frist->InterPreter() * second->InterPreter();
    }
private:
    interpreter* frist;
    interpreter* second;
};

class NumberInterpreter : public interpreter {
public:
    NumberInterpreter(const string s):num(s){}
public:
    int InterPreter() {
        return atoi(num.c_str());
    }
private:
    string num;
};
vector<string> split (string s, string delimiter) {
    size_t pos_start = 0, pos_end, delim_len = delimiter.length();
    string token;
    vector<string> res;

    while ((pos_end = s.find (delimiter, pos_start)) != string::npos) {
        token = s.substr (pos_start, pos_end - pos_start);
        pos_start = pos_end + delim_len;
        res.push_back (token);
    }

    res.push_back (s.substr (pos_start));
    return res;
}
int Parser(string s) {
    vector<string> v = split(s, " ");
    vector<interpreter*> iv;
    return Option(v, iv);
}
int Option(vector<string>& v, vector<interpreter*>& iv) {
    if (iv.size() == 1 && v.size() == 0) {
        return iv[0]->InterPreter();
    }
    if (v.size() == 0) {
        return 0;
    } 

    string n = v[0];
    if (n == "*") {
        if (iv.size() < 2 ) {
            return 0;
        }
        interpreter* second = iv.back();
        iv.pop_back();
        interpreter* frist = iv.back();
        iv.pop_back();

        interpreter* i = new MutilInterpreter(frist, second);
        iv.push_back(i);

    } else if (n == "+") {
        if (iv.size() < 2 ) {
            return 0;
        }
        interpreter* second = iv.back();
        iv.pop_back();
        interpreter* frist = iv.back();
        iv.pop_back();

        interpreter* i = new AddInterpreter(frist, second);
        iv.push_back(i);
    } else {
        interpreter* i = new NumberInterpreter(n);
        iv.push_back(i);
    }
    vector<string> vt;
    for (int i = 1; i < v.size(); ++i) {
        vt.push_back(v[i]);
    }
    return Option(vt, iv);
}
void test_interpreter() {
    cout << "6 100 11 + * 10 * is " << Parser("6 100 11 + * 10 *") << endl;
}
int main() {
    test_interpreter();
    return 0;
}

File: InterPerter_Pattern/go/Interpreter_Pattern.go
package Interpreter_Pattern

import (
	"strconv"
	"strings"
)

type interpreter interface {
	Interpreter() int
}

type MutilInterpreter struct {
	First  interpreter
	Second interpreter
}

func (this *MutilInterpreter) Interpreter() int {
	return this.First.Interpreter() * this.Second.Interpreter()
}

type AddInterpreter struct {
	First  interpreter
	Second interpreter
}

func (this *AddInterpreter) Interpreter() int {
	return this.First.Interpreter() + this.Second.Interpreter()
}

type NumberInterpreter struct {
	Num string
}

func (this *NumberInterpreter) Interpreter() int {
	if n, err := strconv.Atoi(this.Num); err == nil {
		return n
	}
	return 0
}

type Context struct {
	Interpreter []interpreter
}

func (this *Context) Pop() interpreter {
	if len(this.Interpreter) == 0 {
		return nil
	}
	I := this.Interpreter[len(this.Interpreter)-1]
	this.Interpreter = this.Interpreter[0 : len(this.Interpreter)-1]
	return I
}
func (this *Context) Push(I interpreter) {
	this.Interpreter = append(this.Interpreter, I)
}

func Parser(s string) int {
	ctx := Context{}
	ss := strings.Split(s, " ")
	return Option(ss, &ctx)
}

func Option(ss []string, ctx *Context) int {
	if len(ss) == 0 && len(ctx.Interpreter) == 1 {
		return ctx.Pop().Interpreter()
	}
	if len(ss) == 0 {
		return 0
	}
	n := ss[0]
	if n == "*" {
		second, frist := ctx.Pop(), ctx.Pop()
		if second == nil || frist == nil {
			return 0
		}
		i := MutilInterpreter{First: frist, Second: second}
		ctx.Push(&i)
	} else if n == "+" {
		second, frist := ctx.Pop(), ctx.Pop()
		if second == nil || frist == nil {
			return 0
		}
		i := AddInterpreter{First: frist, Second: second}
		ctx.Push(&i)
	} else if IsNumber(n) {
		i := NumberInterpreter{Num: n}
		ctx.Push(&i)
	} else {
		panic("unkonw synxt ")
	}
	return Option(ss[1:], ctx)
}
func IsNumber(s string) bool {
	if len(s) == 0 {
		return false
	}
	for _, n := range s {
		if n > '9' || n < '0' {
			return false
		}
	}
	return true
}
